src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/ReflectionMarker.cs

```diff
@@ -135,40 +145,70 @@ void MarkEvent(in MessageOrigin origin, EventPseudoDesc @event, Origin memberWit
         {
+            if (!_enabled)
+                return;
+
             foreach (var property in type.GetPropertiesOnTypeHierarchy(filter, bindingFlags))
                 MarkProperty(origin, property, memberWithRequirements);
         }
 
         internal void MarkEventsOnTypeHierarchy(in MessageOrigin origin, TypeDesc type, Func<EventPseudoDesc, bool> filter, Origin memberWithRequirements, BindingFlags? bindingFlags = BindingFlags.Default)
         {
+            if (!_enabled)
+                return;
+
             foreach (var @event in type.GetEventsOnTypeHierarchy(filter, bindingFlags))
                 MarkEvent(origin, @event, memberWithRequirements);
         }
 
         internal void MarkStaticConstructor(in MessageOrigin origin, TypeDesc type)
         {
-            if (!type.IsGenericDefinition && !type.ContainsSignatureVariables(treatGenericParameterLikeSignatureVariable: true) && type.HasStaticConstructor)
+            if (!_enabled)
+                return;
+
+            if (!type.IsGenericDefinition && !type.ContainsSignatureVariables(treatGenericParameterLikeSignatureVariable: true) && Factory.PreinitializationManager.HasLazyStaticConstructor(type))
             {
                 // Mark the GC static base - it contains a pointer to the class constructor, but also info
                 // about whether the class constructor already executed and it's what is looked at at runtime.
-                _dependencies.Add(_factory.TypeNonGCStaticsSymbol((MetadataType)type), "RunClassConstructor reference");
+                _dependencies.Add(Factory.TypeNonGCStaticsSymbol((MetadataType)type), "RunClassConstructor reference");
             }
         }
 
-        void WarnOnReflectionAccess(in MessageOrigin origin, TypeSystemEntity entity, Origin memberWithRequirements)
+        void CheckAndWarnOnReflectionAccess(in MessageOrigin origin, TypeSystemEntity entity, Origin memberWithRequirements)
         {
+            if (entity.DoesMemberRequire(DiagnosticUtilities.RequiresUnreferencedCodeAttribute, out CustomAttributeValue<TypeDesc>? requiresAttribute))
+            {
+                if (_typeHierarchyDataFlow)
+                {
+                    _logger.LogWarning(origin, DiagnosticId.DynamicallyAccessedMembersOnTypeReferencesMemberOnBaseWithRequiresUnreferencedCode,
+                        ((TypeOrigin)memberWithRequirements).GetDisplayName(),
+                        entity.GetDisplayName(),
+                        MessageFormat.FormatRequiresAttributeMessageArg(DiagnosticUtilities.GetRequiresAttributeMessage(requiresAttribute.Value)),
+                        MessageFormat.FormatRequiresAttributeUrlArg(DiagnosticUtilities.GetRequiresAttributeUrl(requiresAttribute.Value)));
+                }
+            }
+
```

Comment:
The implementation in linker kind of calls CheckAndReportRequires here. Without it, very basic scenarios in reflection won't warn. For example, I was debugging today in the repro test the following:
``` C#
public static void Main ()
{
	typeof (BaseClassWithRequires).RequiresPublicMethods ();
}

class BaseClassWithRequires
{
	[RequiresUnreferencedCode ("Message")]
	[RequiresAssemblyFiles ("Message")]
	[RequiresDynamicCode ("Message")]
	public virtual void VirtualMethod ()
	{
	}
}
```
Or using the GetMethod() method which has the same effect of marking the method with Requires and not producing the warning.